### PR TITLE
HGCal emulator and unpacker

### PR DESCRIPTION
**Pull request by forthommel**
#### PR description:

This PR introduces an unpacker module+algorithm for HGCal. It provides the missing link between the S-link frames cast into FED raw data and the digis collection already introduced since #38538, and allows future developments for the validation of the whole transmission chain.

As a way to test this unpacking, a full frame emulator is also provided along this PR. It allows the emulation of S-link, capture block, and ECON-D level payloads to be fed to the unpacker to simulate various boundary conditions. It can recast either realistic frames collected in lab/test beam, or fully "trivial" channels content containing the minimal information possible. Along with this emulation, a few metadata objects are introduced for future in-depth validations, when the error handling of the unpacker will be improved.

Various tests, defined along with the DPG (see e.g. [this presentation from 12.04](https://indico.cern.ch/event/1274686/contributions/5353145)) are introduced as "unit tests", and allow these boundary conditions checks to be performed.

A few `const`-qualification of the main data formats were introduced, possibly for the thread safety of some of these tests.

Beside the long development history leading to this PR, it only contains a limited (although >1) number of commits to reflect the various contributions from the experts involved (@pfs, @yulunmiao).

#### PR validation:

This PR is relatively orthogonal to all parts of the validation suite, and only a limited, but successful `runTheMatrix.py` run was performed. As mentioned above, a few unit tests are introduced along with these two components. These can be steered through the standard `scram b runtests`.

#### If this PR is a backport please specify the original PR and why you need to backport that PR. If this PR will be backported please specify to which release cycle the backport is meant for: N/A

### CODE REVIEW

**Comment by civanch** [comment]
please test

File: EventFilter/HGCalRawToDigi/interface/HGCalUnpacker.h [hunk 1]
```diff
@@ -0,0 +1,142 @@
...
+  /// \return vector of HGCROCChannelDataFrame<D>(ID, value)
+  const std::vector<HGCROCChannelDataFrame<D> >& channelData() const { return channelData_; }
+  /// \return vector of 32-bit index, the length is the same as channelData(), link from channel data to the first common mode on ROC (+0,+1,+2,+3 for all four common modes)
+  const std::vector<uint32_t>& commonModeIndex() const { return commonModeIndex_; }
+  /// \return vector of 16-bit common mode data, lowest 10 bits is the ADC of the common mode, padding to 4 for half ROC turned on
+  const std::vector<uint16_t>& commonModeData() const { return commonModeData_; }
+  const std::vector<uint32_t>& badECOND() const { return badECOND_; }
+
+private:
+  const uint32_t erxBodyLeftShift_[16] = {2, 0, 2, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0};
```
Comment (clacaputo) [inline EventFilter/HGCalRawToDigi/interface/HGCalUnpacker.h:114]: Are these numbers (from `L114` to `L132`) something that could change along time? If this is the case, wouldn't be better to add them in a `cfg` file?

Comment (forthommel) [reply EventFilter/HGCalRawToDigi/interface/HGCalUnpacker.h:114]: @clacaputo Does this address your concern about the `const`-ness of these numbers?

Comment (clacaputo) [reply EventFilter/HGCalRawToDigi/interface/HGCalUnpacker.h:114]: @forthommel , it does. @yulunmiao thanks for the clarification.

**Comment by forthommel** [comment]
If only a timeout issue, could we issue a rebuild and see if this fixes it?

**Comment by forthommel** [comment]
@clacaputo is there anything we can do to ease your review of this PR? It's a central piece before future developments on our front.

**Comment by clacaputo** [comment]
+reconstruction

**Comment by forthommel** [comment]
Thanks a lot, that was fast!
@civanch, @mdhildreth, @srimanob, @AdrianoDee, same proposal: if there is anything we can do to ease your review...

**Comment by civanch** [comment]
+1

**Comment by forthommel** [comment]
@AdrianoDee, @srimanob, anything you need from us for your `+upgrade`?

**Comment by pfs** [comment]
all that can be done to speed up the integration of this PR is appreciated - we'll rely on it for the upcoming test beams in august/september. Thanks

**Comment by srimanob** [comment]
@cmsbuild please test

Just to refresh the test before merging.

**Comment by forthommel** [comment]
Good to see this is still running as fine as ~1 month ago :)

**Comment by perrotta** [comment]
+1

**Comment by forthommel** [comment]
Many many thanks to everyone involved!

File: EventFilter/HGCalRawToDigi/plugins/HGCalRawToDigi.cc [hunk 2]
```diff
@@ -0,0 +1,130 @@
...
+    if (fed_data.size() == 0)
+      continue;
+
+    std::vector<uint32_t> data_32bit;
+    auto* ptr = fed_data.data();
+    for (size_t i = 0; i < fed_data.size(); i += 4)
+      data_32bit.emplace_back(((ptr + i) ? ((*(ptr + i) & 0xff) << 0) : 0) +
+                              ((ptr + i + 1) ? ((*(ptr + i + 1) & 0xff) << 8) : 0) +
+                              ((ptr + i + 2) ? ((*(ptr + i + 2) & 0xff) << 16) : 0) +
+                              ((ptr + i + 3) ? ((*(ptr + i + 3) & 0xff) << 24) : 0));
```
Comment (perrotta) [inline EventFilter/HGCalRawToDigi/plugins/HGCalRawToDigi.cc:71]: @forthommel the el8_amd64_gcc12 builds are complaining in the IB about these lines, see e.g. https://cmssdt.cern.ch/SDT/cgi-bin/buildlogs/el8_amd64_gcc12/CMSSW_13_2_X_2023-05-27-1100/EventFilter/HGCalRawToDigi

Indeed: what are you checking here? The pointer always exists, it may point to some non existent field at most.

Could something as follows correspond to what you had in mind, maybe?
```C++
    size_t fed_data_size = fed_data.size();
    for (size_t i = 0; i < fed_data_size; i += 4)
      data_32bit.emplace_back(((*(ptr + i) & 0xff) << 0)  +
                              ((i + 1 < fed_data_size )? (*(ptr + i + 1) & 0xff) << 8) : 0) +
                              ((i + 2 < fed_data_size) ? ((*(ptr + i + 2) & 0xff) << 16) : 0) +
                              ((i + 3 < fed_data_size) ? ((*(ptr + i + 3) & 0xff) << 24) : 0));
```

Comment (forthommel) [reply EventFilter/HGCalRawToDigi/plugins/HGCalRawToDigi.cc:71]: Hi @perrotta !
Thanks for the extra x-check. Sure, this would indeed be a better practice. I can prepare a PR candidate to follow this up.
As a matter of fact I was actually surprised such a simple words merging operation was not already part of the common `FWCore/Utilities` tools!
